Nominate enrollments using `avail_height` on the `enrollment_pool` table

linked0 opened the pull request:
I added the `avail_height` column into the `EnrollmentPool` in order to check if an enrollment can be a validator. If even the enrollment had been promoted as a validator, it can be used as the request to be a validator in the next cycle. So the `getEnrollments` function can select the enrollments which can be a validator for the next block with the `avail_height` and the height passed in as a parameter.

When the `getEnrollments` is called, it uses the `avail_height` column to get the right enrollments at the current block height. The rule is as follows.
1) If the `enrolled_height` of an enrollment is greater than the `avail_height`, we do not return the enrollment.
2) If the current height(the height for which we are nominating minus 1) is smaller than the `avail_height`, we do not return the enrollment.

Relates to #900

linked0 committed:
Add `avail_height` column into `EnrollmentPool`

We add the `avail_height` column into the `EnrollmentPool` in order to check
if an enrollment can be a validator. If even the enrollment had been promoted
as a validator, it can be used as the request to be a validator in the next
cycle. So the `getEnrollments` function can select the enrollments which can
be a validator for next block with the `avail_height` and the height passed
in as a parameter.

linked0 commented:
Ready for review

Geod24 reviewed (changes requested):
I see three issues with this PR:
1) The test code and the implementation are separate. As repeated multiple times, they should come together.
2) The test does not test what it should. If I phrase what the test is doing, in English: "Get the value of a field in the SQL database and check that it is exactly what we put in". That is testing the database, *not* the actual behavior we want to test. What we want to test is that the precondition of #900 are here.
3) What we want is the height at which the enrollment is available, not when the enrollment was inserted. I added [a comment](https://github.com/bpfkorea/agora/issues/900#issuecomment-652717204) in the issue to clarify this.

linked0 commented:
Ready for review.
If this PR is approved today, I can do the following task, or another PR next week to complete the issue #900 . If you review it, I will modify it as soon as possible.

TrustHenry commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -190,6 +190,34 @@ public class EnrollmentManager
+        scope UTXOFinder finder) @safe nothrow
+    {
+        auto enrolled_height = this.getEnrolledHeight(enroll.utxo_key);
+        ulong avail_height;
```
Annotation for avail_height is required.

linked0 replied:
> Annotation for avail_height is required.

It would be better to describe the purpose of the `avail_height`. Thanks.

linked0 replied:
> Annotation for avail_height is required.

It would be better to describe the purpose of the avail_height. Thanks.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -190,6 +190,34 @@ public class EnrollmentManager
+        auto enrolled_height = this.getEnrolledHeight(enroll.utxo_key);
+        ulong avail_height;
+
+        if (enrolled_height == ulong.max)
+            avail_height = block_height + 1 + this.params.ValidatorCycle;
```
I don't understand this part. Why do you set `avail_height` to 1009 blocks in the future when there is no Enrollment ?

linked0 replied:
The `block_height` + 1 means the enrolled height if the enrollment has been enrolled at the next block of the `block_height`. So the `avail_height` becomes the last height at which a node plays a role as a validator. This assignment for the `avail_height` is only relevant to nodes requesting enrollment after the Genesis block.

Geod24 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -84,35 +87,42 @@ public class EnrollmentPool
-                this.db.execute("INSERT INTO enrollment_pool (key, val) VALUES (?, ?)",
-                    enroll.utxo_key.toString(), buffer);
-            }();
+            if (this.hasEnrollment(enroll.utxo_key))
+            {
+                () @trusted {
+                    this.db.execute("UPDATE enrollment_pool SET " ~
+                        "avail_height = ? WHERE key = ?",
+                        avail_height.value, enroll.utxo_key.toString());
+                }();
+                return false;
+            }
+            else
+            {
+                static ubyte[] buffer;
+                try
+                {
+                    serializeToBuffer(enroll, buffer);
+                }
+                catch (Exception ex)
+                {
+                    log.error("Serialization error: {}, Data was: {}", ex.msg,
+                        enroll);
+                    return false;
+                }
+                () @trusted {
+                    this.db.execute("INSERT INTO enrollment_pool " ~
+                        "(key, val, avail_height) VALUES (?, ?, ?)",
+                        enroll.utxo_key.toString(), buffer,
+                        avail_height.value);
+                }();
+            }
```
https://www.sqlitetutorial.net/sqlite-replace-statement/

linked0 replied:
> https://www.sqlitetutorial.net/sqlite-replace-statement/

I will apply it to the code. Thanks.

linked0 commented:
Ready for review. Please!

linked0 commented:
Ready for review.

Geod24 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -169,19 +175,25 @@ public class EnrollmentPool
+            auto results = this.db.execute("SELECT avail_height FROM " ~
+                "enrollment_pool WHERE key = ?", enroll_hash.toString());
+            if (!results.empty &&
+                Height(results.oneValue!(size_t)) == avail_height)
```
You don't need the `Height()` here

linked0 replied:
This needs `Height()` because of the below code. Let's talk about that IRL.
```
if (this.enroll_man.addEnrollment(enroll, this.ledger.getBlockHeight(),
            this.utxo_set.getUTXOFinder()))
{
    this.network.sendEnrollment(enroll);
}
```

Geod24 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -169,19 +175,25 @@ public class EnrollmentPool
-            return results.front().peek!bool(0);
+            auto results = this.db.execute("SELECT avail_height FROM " ~
+                "enrollment_pool WHERE key = ?", enroll_hash.toString());
+            if (!results.empty &&
+                Height(results.oneValue!(size_t)) == avail_height)
+                return true;
+            else
+                return false;
```
Why would you write:
```D
if (condition)
    return true;
else
    return false;
```

Instead of simply `return condition;` ? I feel that this pattern is repeated way too frequently, please remember the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle).

Geod24 reviewed (changes requested):
There are too many things wrong here. I noted a few of them but the list is not exhaustive.
Let's discuss that IRL on Tuesday.

Geod24 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -169,19 +175,25 @@ public class EnrollmentPool
-            auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM " ~
-                "enrollment_pool WHERE key = ?)", enroll_hash.toString());
-            return results.front().peek!bool(0);
+            auto results = this.db.execute("SELECT avail_height FROM " ~
+                "enrollment_pool WHERE key = ?", enroll_hash.toString());
```
```suggestion
            auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM " ~
                "enrollment_pool WHERE key = ? AND avail_height = ?", enroll_hash.toString(), avail_heigth);
```

linked0 committed:
Modify the `getEnrollments` to use `avail_height`

When the `getEnrollment` is called, it uses the `avail_height` column to get
right enrollments at current block height. The rule is as follows.
1) If the `enrolled_height` of an enrollment is greater than the `avail_height`
we do not return the enrollment.
2) If the current height(the height for which we are nominating minus 1) is
smaller than the `avail_height`, we do not return the enrollment.

linked0 commented:
Ready for review.
I added an additional commit for the `getEnrollments` function.
I will make additional PR needed today.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -190,6 +190,36 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    public bool addEnrollment (const ref Enrollment enroll, Height block_height,
```
If this function should always be used instead of `pool.add`, then we need to remove the `pool()` getter in this class.

linked0 replied:
> If this function should always be used instead of `pool.add`, then we need to remove the `pool()` getter in this class.

The `ValdiatorSet` and the `EnrollmentPool` are needed at the same time for the operation, so it might be better for the `EnrollmentManager` to have the function. I'm not sure which is better. Do you think putting the code in the Ledger would be better? Anyway, if removing the pool() getter is needed, I'll do it.

AndrejMitrovic replied:
I'm saying what if the code calls `pool.add` instead of `EnrollmentManager.addEnrollment` by mistake? There is additional logic now in the enrollment manager which is not in the pool.

linked0 replied:
> I'm saying what if the code calls `pool.add` instead of `EnrollmentManager.addEnrollment` by mistake? There is additional logic now in the enrollment manager which is not in the pool.

Right, the `pool()` getter will be removed.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -241,12 +241,16 @@ public class EnrollmentManager
 
         static Enrollment[] pool_enrolls;
-        this.enroll_pool.getEnrollments(pool_enrolls);
+        this.enroll_pool.getEnrollments(pool_enrolls, height);
```
Hmm shouldn't this be `height + 1`?

AndrejMitrovic replied:
Because the `height` passed here is the already externalized block height, we want to know if the enrollment is available for the next block right?

linked0 replied:
> Because the `height` passed here is the already externalized block height, we want to know if the enrollment is available for the next block right?

The meaning of the parameter is the current block height and the function finds the enrollments that are available for the next height. But it can be modified as you suggested in order to reduce confusion.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1054,3 +1058,59 @@ unittest
+
+    // create an EnrollmentManager
+    const validator_cycle = 20;
+    auto man = new EnrollmentManager(":memory:", key_pair,
```
```suggestion
    scope man = new EnrollmentManager(":memory:", key_pair,
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -237,20 +237,54 @@ public class EnrollmentPool
+                return Height(0);
+
+            return Height(results.oneValue!(size_t));
+
```
```suggestion
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -237,20 +237,54 @@ public class EnrollmentPool
+            enroll_hash = key for the available block height
+
+        Returns:
+            the available block height
```
```suggestion
            the available block height, or 0 if the enrollment isn't found
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -237,20 +237,54 @@ public class EnrollmentPool
+
     /***************************************************************************
 
         Get the unregistered enrollments in the block
```
This doc should be updated:

Get the unregistered enrollments from the pool

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -190,6 +190,36 @@ public class EnrollmentManager
+    {
+        auto enrolled_height = this.getEnrolledHeight(enroll.utxo_key);
+
+        // The last height at which a node plays a role as a validator.
```
The comment should be corrected.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -211,12 +241,16 @@ public class EnrollmentManager
+        this.enroll_pool.getEnrollments(pool_enrolls, height);
         foreach (enroll; pool_enrolls)
         {
             const enroll_height = this.getEnrolledHeight(enroll.utxo_key);
+            const avail_height =
+                this.enroll_pool.getAvailableHeight(enroll.utxo_key);
```
Why would we need `getAvailableHeight` if we know `getEnrollments` already filters them out ???

linked0 replied:
> Why would we need getAvailableHeight if we know getEnrollments already filters them out ???

It's because we need to filter out the enrollment which left unintentionally in the pool although it had become a validator.

linked0 committed:
Remove the `pool` function from `EnrollmentManager`

The addEnrollment is created in the previous commit, so we should
remove the `pool` function because the `add` function can called
mistakenly. And we add the `getEnrollment` and the `removeEnrollment`
as a result of it.

linked0 commented:
Ready for review.
I applied the suggestions, and removed the `pool` function from the `EnrollmentManager`.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1025,12 +1025,13 @@ unittest
     foreach (idx, hash; utxo_hashes)
     {
-        assert(ledger.enroll_man.pool.getEnrollment(hash, stored_enroll));
+        assert((stored_enroll = ledger.enroll_man.getEnrollment(hash)) !=
+            Enrollment.init);
```
Please break in two lines

Geod24 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -201,40 +201,39 @@ public class EnrollmentPool
             log.error("Exception occured on getEnrollment: {}, " ~
                 "Key for enrollment: {}", ex.msg, enroll_hash);
-            return false;
+            return enroll;
```
Returning `Enrollment.init` will be more explicit.

Geod24 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -201,40 +201,39 @@ public class EnrollmentPool
         }
 
-        return false;
+        return enroll;
```
Same comment about `init` here

linked0 commented:
I applied the suggestions. Thanks.

MichaelKim20 commented on source/agora/consensus/EnrollmentPool.d:
```diff
@@ -239,18 +239,51 @@ public class EnrollmentPool
+            if (results.empty)
+                return Height(0);
+
+            return Height(results.oneValue!(size_t));
```
When it has multiple records, it will return a record with the smallest 'avail_height'?
However, 'avail_height' is not a primary key, so it should be checked.

linked0 replied:
> When it has multiple records, it will return a record with the smallest 'avail_height'?
> However, 'avail_height' is not a primary key, so it should be checked.

The `enroll_hash` which is a UTXO hash is a primary key for the table, so the `avail_height` for the key is only one. Thanks.

MichaelKim20 replied:
Oh, I see. We can always register only one data in advance.

MichaelKim20 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -241,12 +241,16 @@ public class EnrollmentManager
+            assert(avail_height != Height(0));
             if (enroll_height == ulong.max ||
-                height >= enroll_height + this.params.ValidatorCycle - 1)
+                (avail_height >= enroll_height &&
+                    height >= enroll_height + this.params.ValidatorCycle - 1))
```
The person who enrolls the information will feel bored because the information he or she has enrolled is not contained in the block and there is no change.

linked0 replied:
The code is for the node which is willing to re-enroll as a validator although it is a validator. The code means as follows in fact.
`(enrolled_height != ulong.max && avail_height >= enroll_height && height >= enroll_height + this.params.ValidatorCycle - 1))`

If the `enrolled_height` equals to `ulong.max`, it will be contained in a block probably because it is a candidate for a validator.
Thanks.
